File: Lecture11Code/PrimalityTest.pl
#!/usr/bin/env perl

##  PrimalityTest.pl
##  Author: Avi Kak
##  Date:   February 28, 2016

##  An implementation of the Miller-Rabin primality test

### You can call this script with either no comamnd-line args or with just one
### command-line arg.  If you call it with no args, it returns primality results on a
### set of randomly altered 36 primes.  On the other hand, if you call it with just
### one arg, it returns the answer for that integer.

use strict;
use warnings;

unless (@ARGV) {
    my @primes = qw[ 179 233 283 353 419 467 547 607 661 739 811 877
                     947 1019 1087 1153 1229 1297 1381 1453 1523 1597
                     1663 1741 1823 1901 7001 7109 7211 7307 7417 7507
                     7573 7649 7727 7841 ];                                  #(M1)
    foreach my $p (@primes) {                                                #(M2)
        $p += 1 + int(rand(10));                                             #(M3)
        my $probability_of_prime = test_integer_for_prime($p);               #(M4)
        $probability_of_prime > 0 ?                                          #(M5)
          print "$p is prime with probability: $probability_of_prime\n" :    #(M6)
          print "$p is composite\n";                                         #(M7)         
    }
} elsif (@ARGV == 1) {                                                       #(M8)
    my $p = shift;                                                           #(M9)
    die "Your number is too large for this script. Instead, try the " .
        "script 'PrimalityTestWithBigInt.pl'\n"
        if $p > 0x7f_ff_ff_ff;                                               #(M10)
    my $probability_of_prime = test_integer_for_prime($p);                   #(M11)
    $probability_of_prime > 0 ?                                              
      print "$p is prime with probability: $probability_of_prime\n" :   
      print "$p is composite\n";                                             #(M12)          
} else {                                                                     #(M13)
    die "You cannot call 'PrimalityTest.py' with more " .             
        "than one command-line argument";                                    #(M14)
}

sub test_integer_for_prime {                                                 #(A1)
    my $p = shift;                                                           #(A2)
    return 0 if $p == 1;                                                     #(A3)
    my @probes = (2,3,5,7,11,13,17);                                         #(A4)
    my @in_probes = grep {$p == $_} @probes;                                 #(A5)
    return 1 if @in_probes;                                                  #(A6)
    my $p_mod_a = 1;                                                         #(A7)
    map { $p_mod_a = 0 if $p % $_ == 0 } @probes;                            #(A8)
    return 0 if $p_mod_a == 0;                                               #(A9)
    my ($k, $q) = (0, $p - 1);                                               #(A10)
    while (! ($q & 1)) {                                                     #(A11)
        $q >>= 1;                                                            #(A12)
        $k += 1;                                                             #(A13)
    }                                                         
    my ($a_raised_to_q, $a_raised_to_jq, $primeflag);                        #(A14)
    foreach my $a (@probes) {                                                #(A15)
        my ($base,$exponent) = ($a,$q);                                      #(A16)
        my $a_raised_to_q = 1;                                               #(A17)
        while ((int($exponent) > 0)) {                                       #(A18)
            $a_raised_to_q = ($a_raised_to_q * $base) % $p 
                                              if int($exponent) & 1;         #(A19)
            $exponent = $exponent >> 1;                                      #(A20)
            $base = ($base * $base) % $p;                                    #(A21)
        }
        next if $a_raised_to_q == 1;                                         #(A22)
        next if ($a_raised_to_q == ($p - 1)) && ($k > 0);                    #(A23)
        $a_raised_to_jq = $a_raised_to_q;                                    #(A24)
        $primeflag = 0;                                                      #(A25)
        foreach my $j (0 .. $k - 2) {                                        #(A26)
            $a_raised_to_jq = ($a_raised_to_jq ** 2) % $p;                   #(A27)
            if ($a_raised_to_jq == $p-1) {                                   #(A28)
                $primeflag = 1;                                              #(A29)
                last;                                                        #(A30)
            }
        }
        return 0 if ! $primeflag;                                            #(A31)
    }
    my $probability_of_prime = 1 - 1.0/(4 ** scalar(@probes));               #(A32)
    return $probability_of_prime;                                            #(A33)
}

File: Lecture11Code/PrimalityTest.py
#!/usr/bin/env python

##  PrimalityTest.py
##  Author: Avi Kak
##  Date:   February 18, 2011
##  Updated:  February 28, 2016
##  An implementation of the Miller-Rabin primality test

### You can call this script with either no comamnd-line args or with just one
### command-line arg.  If you call it with no args, it returns primality results on a
### set of randomly altered 36 primes.  On the other hand, if you call it with just
### one arg, it returns the answer for that integer.

def test_integer_for_prime(p):                                               #(A1)
    if p == 1: return 0                                                      #(A2)
    probes = [2,3,5,7,11,13,17]                                              #(A3)
    if p in probes: return 1                                                 #(A4)
    if any([p % a == 0 for a in probes]): return 0                           #(A5)
    k, q = 0, p-1        # need to represent p-1 as  q * 2^k                 #(A6)
    while not q&1:                                                           #(A7)
        q >>= 1                                                              #(A8)
        k += 1                                                               #(A9)
    for a in probes:                                                         #(A10)
        a_raised_to_q = pow(a, q, p)                                         #(A11)
        if a_raised_to_q == 1: continue                                      #(A12)
        if (a_raised_to_q == p-1) and (k > 0): continue                      #(A13)
        a_raised_to_jq = a_raised_to_q                                       #(A14)
        primeflag = 0                                                        #(A15)
        for j in range(k-1):                                                 #(A16)
            a_raised_to_jq = pow(a_raised_to_jq, 2, p)                       #(A17)
            if a_raised_to_jq == p-1:                                        #(A18)
                primeflag = 1                                                #(A19)
                break                                                        #(A20)
        if not primeflag: return 0                                           #(A21)
    probability_of_prime = 1 - 1.0/(4 ** len(probes))                        #(A22)
    return probability_of_prime                                              #(A23)

primes = [179, 233, 283, 353, 419, 467, 547, 607, 661, 739, 811, 877, \
          947, 1019, 1087, 1153, 1229, 1297, 1381, 1453, 1523, 1597, \
          1663, 1741, 1823, 1901, 7001, 7109, 7211, 7307, 7417, 7507, \
          7573, 7649, 7727, 7841]                                            #(A24)

if __name__ == '__main__':

    import sys                                                               #(M1)
    import random                                                            #(M2)

    if len(sys.argv) == 1:                                                   #(M3)
        for p in primes:                                                     #(M4)
            p += random.randint(1,10)                                        #(M5)
            probability_of_prime = test_integer_for_prime(p)                 #(M6)
            if probability_of_prime > 0:                                     #(M7)
                print("%d is prime with probability: %f" %(p,probability_of_prime))
                                                                             #(M8)
            else:                                                            #(M9)
                print("%d is composite" % p)                                 #(M10)
    elif len(sys.argv) == 2:                                                 #(M11)
        p = int(sys.argv[1])                                                 #(M12)
        probability_of_prime = test_integer_for_prime(p)                     #(M13)             
        if probability_of_prime > 0:                                         #(M14)
            print("%d is prime with probability: %f" %(p,probability_of_prime))
                                                                             #(M15)
        else:                                                                #(M16)
            print("%d is composite" % p)                                     #(M17)
    else:                                                                    #(M18)
        sys.exit("""You cannot call 'PrimalityTest.py' with more """         #(M19)
                 """than one command-line argument""")
